FlowCrypt/src/main/java/com/flowcrypt/email/jetpack/workmanager/HandlePasswordProtectedMsgWorker.kt

```diff
@@ -0,0 +1,424 @@
+
+      return@withContext rescheduleIfActiveAccountWasChanged(account)
+    } catch (e: Exception) {
+      e.printStackTrace()
+      return@withContext Result.failure()
+    } finally {
+      LogsUtil.d(TAG, "work was finished")
+    }
+  }
+
+  private suspend fun prepareAndUploadPasswordProtectedMsgsToFES(account: AccountEntity) =
+    withContext(Dispatchers.IO) {
+      val apiRepository = FlowcryptApiRepository()
+      val keysStorage = KeysStorageImpl.getInstance(applicationContext)
+      val accountSecretKeys = PGPSecretKeyRingCollection(keysStorage.getPGPSecretKeyRings())
+
+      var list: List<MessageEntity>
+      var lastMsgUID = 0L
+      while (true) {
+        list = roomDatabase.msgDao().getOutboxMsgsByStatesSuspend(
+          account = account.email,
+          msgStates = listOf(MessageState.NEW_PASSWORD_PROTECTED.value)
+        )
+
+        if (list.isEmpty()) break
+        val msgEntity = list.firstOrNull { it.uid > lastMsgUID } ?: list[0]
+
+        lastMsgUID = msgEntity.uid
+
+        try {
+          if (msgEntity.isEncrypted == true && msgEntity.isPasswordProtected) {
+            //get msg attachments(including forwarded that can be encrypted)
+            val attachments = getAttachments(msgEntity)
+
+            //create MimeMessage from content + attachments
+            val plainMimeMsgWithAttachments = getMimeMessage(account, msgEntity, attachments)
+
+            //get recipients that will be used to create password-encrypted msg
+            val toCandidates = getRecipients(plainMimeMsgWithAttachments, Message.RecipientType.TO)
+            val ccCandidates = getRecipients(plainMimeMsgWithAttachments, Message.RecipientType.CC)
+            val bccCandidates =
+              getRecipients(plainMimeMsgWithAttachments, Message.RecipientType.BCC)
+
+            if (toCandidates.isEmpty() && ccCandidates.isEmpty() && bccCandidates.isEmpty()) {
+              throw IllegalStateException("Wrong password-protected implementation")
+            }
+
+            //start of creating and uploading a password-protected msg to FES
+            val fromAddress = plainMimeMsgWithAttachments.getFromAddress()
+            val domain = EmailUtil.getDomain(fromAddress)
+            val idToken = getGoogleIdToken()
+            val replyToken = fetchReplyToken(apiRepository, domain, idToken)
+            val messageUploadRequest = MessageUploadRequest(
+              associateReplyToken = replyToken,
+              from = fromAddress,
+              to = toCandidates.map { (it as InternetAddress).address },
+              cc = ccCandidates.map { (it as InternetAddress).address },
+              bcc = bccCandidates.map { (it as InternetAddress).address }
+            )
+
+            //prepare bodyWithReplyToken
+            val replyInfo = Base64.getEncoder().encodeToString(
+              GsonBuilder().create().toJson(messageUploadRequest).toByteArray()
+            )
+
+            val infoDiv = genInfoDiv(replyInfo)
```

Comment:
I recommend that `genInfoDiv` should take only `messageUploadRequest` as parameter, and generate `replyInfo` inside the method (and therefore less clutter here in the main method of the file). While doing this, the format of the `replyInfo` needs to be changed as per #1636

Reply:
Will be resolved in #1636